src/polar/kernels.jl

```diff
@@ -113,6 +113,95 @@ function residual_polar!(F, v_m, v_a,
     wait(ev)
 end
 
+KA.@kernel function residual_adj_kernel!(F, adj_F, v_m, adj_v_m, v_a, adj_v_a,
+                                  ybus_re_nzval, ybus_re_colptr, ybus_re_rowval,
+                                  ybus_im_nzval, ybus_im_colptr, ybus_im_rowval,
+                                  pinj, adj_pinj, qinj, adj_qinj, pv, pq, nbus)
+
+    npv = size(pv, 1)
+    npq = size(pq, 1)
+
+    i = @index(Global, Linear)
+    # REAL PV: 1:npv
+    # REAL PQ: (npv+1:npv+npq)
+    # IMAG PQ: (npv+npq+1:npv+2npq)
+    fr = (i <= npv) ? pv[i] : pq[i - npv]
+    F[i] -= pinj[fr]
+    if i > npv
+        F[i + npq] -= qinj[fr]
+    end
+    @inbounds for c in ybus_re_colptr[fr]:ybus_re_colptr[fr+1]-1
+        # Forward loop
+        to = ybus_re_rowval[c]
+        aij = v_a[fr] - v_a[to]
+        # f_re = a * cos + b * sin
+        # f_im = a * sin - b * cos
+        coef_cos = v_m[fr]*v_m[to]*ybus_re_nzval[c]
+        coef_sin = v_m[fr]*v_m[to]*ybus_im_nzval[c]
+
+        cos_val = cos(aij)
+        sin_val = sin(aij)
+        F[i] += coef_cos * cos_val + coef_sin * sin_val
+        if i > npv
+            F[npq + i] += coef_cos * sin_val - coef_sin * cos_val
+        end
+        if i > npv
+            F[npq + i] += coef_cos * sin_val - coef_sin * cos_val
+        end
+        # Reverse loop
+        adj_coef_cos = 0.0
+        adj_coef_sin = 0.0
+        adj_cos_val  = 0.0
+        adj_sin_val  = 0.0
+
+        if i > npv
+            adj_coef_cos +=  sin_val  * adj_F[npq + i]
+            adj_coef_sin += -cos_val  * adj_F[npq + i]
+            adj_cos_val  += -coef_sin * adj_F[npq + i]
+            adj_sin_val  +=  coef_cos * adj_F[npq + i]
+        end
+
+        adj_coef_cos +=  cos_val  * adj_F[i]
+        adj_coef_sin +=  sin_val  * adj_F[i]
+        adj_cos_val  +=  coef_cos * adj_F[i]
+        adj_sin_val  +=  coef_sin * adj_F[i]
+
+        adj_aij =   cos(aij)*adj_sin_val
+        adj_aij += -sin(aij)*adj_cos_val
+
+        adj_v_m[fr] += v_m[to]*ybus_im_nzval[c]*adj_coef_sin
+        adj_v_m[to] += v_m[fr]*ybus_im_nzval[c]*adj_coef_sin
+        adj_v_m[fr] += v_m[to]*ybus_re_nzval[c]*adj_coef_cos
+        adj_v_m[to] += v_m[fr]*ybus_re_nzval[c]*adj_coef_cos
+
+        adj_v_a[fr] += adj_aij
+        adj_v_a[to] -= adj_aij
+    end
+    # TODO: LHS is not always a dual
+    # if i > npv
+    #     adj_qinj[fr] -= adj_F[i + npq]
+    # end
+    # adj_pinj[fr] -= adj_F[i]
+end
+
+function residual_adj_polar!(F, adj_F, v_m, adj_v_m, v_a, adj_v_a,
+                         ybus_re, ybus_im,
+                         pinj, adj_pinj, qinj, adj_qinj, pv, pq, nbus)
+    npv = length(pv)
+    npq = length(pq)
+    if isa(F, Array)
+        kernel! = residual_adj_kernel!(KA.CPU(), 4)
```

Comment:
Why using 4 CPU threads here?

Reply:
No idea. That 4 must have emerged at some point.